chirp/drivers/retevis_ha1g.py

```diff
@@ -0,0 +1,2111 @@
+    page_count = 1
+    page_index = 0
+    item_bytes = b""
+    max_retries = 5
+    retry_delay = 0.05
+    while page_index < page_count:
+        success = False
+        for attempt in range(max_retries):
+            flag, new_data_bytes = handle_connect_ver(
+                Get_ReadItemBytes(item, page_index),
+                serial, self.read_page_len)
+            if (not flag
+               or (len(new_data_bytes) > 14 and new_data_bytes[14] != item)):
+                time.sleep(retry_delay)
+                continue
+            if page_index == 0:
+                page_count = new_data_bytes[18] | (new_data_bytes[19] << 8)
+            status.cur += len(new_data_bytes)
+            self.status_fn(status)
+            item_bytes += new_data_bytes[20:-3]
+            page_index += 1
+            success = True
+            break
+        if not success:
+            raise errors.RadioError("radio reported failure")
+    return item_bytes
+
+
+def write_item_current_page_bytes(self, serial, item_Bytes: bytes,
+                                  item: int, status):
+    if not item_Bytes:
+        raise ValueError("item_bytes cannot be empty")
+    max_retries = 5
+    retry_delay = 0.05
+    page_count = math.ceil(len(item_Bytes) / self.write_page_len)
+    if page_count == 0:
+        return True
+    for i in range(0, page_count):
+        for attempt in range(max_retries):
+            current_page_bytes = get_write_item_send_bytes(
+                item, i, page_count,
+                item_Bytes[
+                    i * self.write_page_len: (i + 1) * self.write_page_len])
+            flag, new_data_bytes = handle_connect_ver(
+                current_page_bytes,
+                serial, self.read_page_len)
+            if (not flag
+               or (len(new_data_bytes) > 14 and new_data_bytes[14] != item)):
+                time.sleep(retry_delay)
+                continue
+            status.cur += len(current_page_bytes)
+            self.status_fn(status)
+            break
+
+
+def get_write_item_bytes(all_bytes: bytearray, item):
+    """
+       Get the bytes for a specific memory region
+       - Uses MEMORY_REGIONS_RANGES to find start and length
+    """
+    if item not in MEMORY_REGIONS_RANGES:
+        LOG.debug(f"Unknown memory Range:{item}")
+        return b""
+    start, length = MEMORY_REGIONS_RANGES[item]
+    return all_bytes[start:start+length]
+
+
+def read_item_handle_connect_event(all_bytes: bytearray,
+                                   item_bytes: bytes, item):
+    """
+    Put the given region bytes into the main memory array
+    - Uses MEMORY_REGIONS_RANGES to find start and length
+    """
+    if item not in MEMORY_REGIONS_RANGES:
+        LOG.debug(f"Unknown memory Range:{item}")
+    start, length = MEMORY_REGIONS_RANGES[item]
+    item_len = min(len(item_bytes), length)
+    all_bytes[start:start + item_len] = item_bytes[:item_len]
```

Comment:
This is modifying the caller's copy of this array. Best to ether modify it (and not return it) or make a copy and return the modified copy.